internal/chain/reward.go

```diff
@@ -2,133 +2,302 @@ package chain
 		}
 	}
+	return payable, missing, nil
+}
 
-	// do it quickly
-	w := sync.WaitGroup{}
-	for i := range missing {
-		w.Add(1)
-		go func(member int) {
-			defer w.Done()
-			// Slash if missing block was long enough
-			if slasher.ShouldSlash(missing[member].BlsPublicKey) {
-				// TODO Logic
-			}
-		}(i)
+func ballotResult(
+	bc engine.ChainReader, header *block.Header, shardID uint32,
+) (shard.SlotList, shard.SlotList, shard.SlotList, error) {
+	// TODO ek – retrieving by parent number (blockNum - 1) doesn't work,
+	//  while it is okay with hash.  Sounds like DB inconsistency.
+	//  Figure out why.
+	parentHeader := bc.GetHeaderByHash(header.ParentHash())
+	if parentHeader == nil {
+		return nil, nil, nil, ctxerror.New(
+			"cannot find parent block header in DB",
+			"parentHash", header.ParentHash(),
+		)
+	}
+	parentShardState, err := bc.ReadShardState(parentHeader.Epoch())
+	if err != nil {
+		return nil, nil, nil, ctxerror.New(
+			"cannot read shard state", "epoch", parentHeader.Epoch(),
+		).WithCause(err)
+	}
+	parentCommittee := parentShardState.FindCommitteeByID(shardID)
+
+	if parentCommittee == nil {
+		return nil, nil, nil, ctxerror.New(
+			"cannot find shard in the shard state",
+			"parentBlockNumber", parentHeader.Number(),
+			"shardID", parentHeader.ShardID(),
+		)
 	}
 
-	w.Wait()
+	payable, missing, err := blockSigners(header, parentCommittee)
+	return parentCommittee.Slots, payable, missing, err
+}
+
+func ballotResultBeaconchain(
+	bc engine.ChainReader, header *block.Header,
+) (shard.SlotList, shard.SlotList, shard.SlotList, error) {
+	return ballotResult(bc, header, shard.BeaconChainShardID)
+}
+
+// AccumulateRewards credits the coinbase of the given block with the mining
+// reward. The total reward consists of the static block reward and rewards for
+// included uncles. The coinbase of each uncle block is also rewarded.
+func AccumulateRewards(
+	bc engine.ChainReader, state *state.DB, header *block.Header,
+	rewarder reward.Distributor, slasher slash.Slasher,
+	beaconChain engine.ChainReader,
+) error {
+
+	blockNum := header.Number().Uint64()
+	if blockNum == 0 {
+		// genesis block has no parent to reward.
+		return nil
+	}
+
+	if bc.Config().IsStaking(header.Epoch()) &&
+		bc.CurrentHeader().ShardID() != shard.BeaconChainShardID {
+		return nil
+	}
+
+	if bc.Config().IsStaking(header.Epoch()) &&
+		bc.CurrentHeader().ShardID() == shard.BeaconChainShardID {
+
+		// Take care of my own beacon chain committee, _ is missing, for slashing
+		members, payable, _, err := ballotResultBeaconchain(beaconChain, header)
+		if err != nil {
+			return err
+		}
+
+		votingPower := votepower.Compute(members)
+
+		for beaconMember := range payable {
+			// TODO Give out whatever leftover to the last voter/handle
+			// what to do about share of those that didn't sign
+			voter := votingPower.Voters[payable[beaconMember].BlsPublicKey]
+			if !voter.IsHarmonyNode {
+				due := BlockRewardStakedCase.Mul(voter.EffectivePercent)
```

Comment:
EffectivePercent is only the percent within 32%, right? Then we are only giving out at most 32% of the total block rewards.